File: context_os_runtime/memory_router.py
from __future__ import annotations

from pathlib import Path

from pydantic import BaseModel

from .models import ProjectManifest


class MemoryRoute(BaseModel):
    project_namespace: str
    project_db_path: Path
    global_db_path: Path
    global_memory_read: bool
    global_memory_write: bool


def build_memory_route(
    manifest: ProjectManifest,
    repo_root: Path,
    global_root: Path,
) -> MemoryRoute:
    return MemoryRoute(
        project_namespace=manifest.memory_namespace,
        project_db_path=repo_root / "data_store" / "knowledge.db",
        global_db_path=global_root / "knowledge.db",
        global_memory_read=manifest.global_memory_read,
        global_memory_write=manifest.global_memory_write,
    )

File: src/core/memory-router.ts
import { join } from 'node:path';

export type Scope = 'project' | 'global';

export interface MemoryRoute {
  project_namespace: string;
  project_db_path: string;
  global_db_path: string;
  global_memory_read: boolean;
  global_memory_write: boolean;
}

export interface MemoryManifestSubset {
  memory_namespace: string;
  global_memory_read?: boolean;
  global_memory_write?: boolean;
}

export function buildMemoryRoute(args: {
  manifest: MemoryManifestSubset;
  repoRoot: string;
  globalRoot: string;
}): MemoryRoute {
  return {
    project_namespace: args.manifest.memory_namespace,
    project_db_path: join(args.repoRoot, 'data_store', 'knowledge.db'),
    global_db_path: join(args.globalRoot, 'knowledge.db'),
    global_memory_read: args.manifest.global_memory_read ?? true,
    global_memory_write: args.manifest.global_memory_write ?? false,
  };
}

export function scopeToDbPath(route: MemoryRoute, scope: Scope): string {
  return scope === 'global' ? route.global_db_path : route.project_db_path;
}
